docs/sources/clients/promtail/logrotation.md

```diff
@@ -0,0 +1,137 @@
+# Promtail and Log Rotation
+
+## Why does log rotation matters?
+
+At any point in time, there may be three processes working on a log file as shown in the image below.
+![block_diagram](./logrotation-components.png)
+
+1. Appender - A writer that keeps appending to a log file. This can be your application or some system daemons like Syslog, Docker log driver or Kubelet, etc.
+2. Tailer - A reader that reads log lines as they are appended e.g agents like Promtail.
+3. Log Rotater - A process that rotates the log file either based on time (say scheduled every day) or size (say a log file reached its max size).
+
+> **NOTE:** Here `fd` defines a file descriptor. Once a file is open for read or write, OS gives back a unique file descriptor (usually an integer) per process, and all the operations like read and write are done over that file descriptor. In other words, once the file is opened successfully, the file descriptor matters more than the file name.
+
+One of the critical components here is the log rotater. Let's understand how it impacts other components like the appender and tailer.
+
+In general, when a log is rotated, it can be done in two ways.
+
+Given a log file `error.log`
+
+1. Copy the log file with a different name e.g `error.log.1` and truncate the original log file `error.log`.
+2. Rename the log file with a different name e.g: `error.log.1` and create a new log file with the original name `error.log`.
+
+In both cases, after log rotation, all new log lines go to the original `error.log` file.
+
+These types are shown in the images below.
+
+### Copy and Truncate
+![block_diagram](./logrotation-copy-and-truncate.png)
+
+### Rename and Create
+![block_diagram](./logrotation-rename-and-create.png)
+
+Both types of log rotation seem to give the same result. However, there are some subtle differences.
+
+(1) favors the `appender` as its descriptor for the original log file `error.log` doesn't change. Therefore, it can keep writing to the same file descriptor. In other words, re-opening the file is not needed.
+
+However (1) has a serious problem if we bring the `tailer` into account. There is a race between truncating the file and the `tailer` finishing reading that log file. Meaning, there is a high chance of the log rotation mechanism truncating the file `error.log` before the `tailer` reads everything from it.
+
+This is where (2) can help. Here, when the log file `error.log` is renamed to `error.log.1`, the `tailer` still holds the file descriptor of `error.log.1`. Therefore, it can continue reading the log file until it's completed. But that comes with the tradeoff: with (2), you have to signal the `appender` to reopen `error.log` (and `appender` should be able to reopen it). Otherwise, it would keep writing to `error.log.1` as the file descriptor won't change. The good news is that most of the popular `appender` solutions (e.g: Syslog, Kubelet, Docker log driver) support reopening log files when they are renamed.
```

Comment:
```suggestion
This is where Rename and Create(2) can help. Here, when the log file `error.log` is renamed to `error.log.1`, the `tailer` still holds the file descriptor of `error.log.1`. Therefore, it can continue reading the log file until it is completed. But that comes with the tradeoff: with (2), you have to signal the `appender` to reopen `error.log` (and `appender` should be able to reopen it). Otherwise, it would keep writing to `error.log.1` as the file descriptor won't change. The good news is that most of the popular `appender` solutions (for example, Syslog, Kubelet, Docker log driver) support reopening log files when they are renamed.
```